File: security/encryption.go
package security

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
)

type FieldEncryptor struct {
	secretKey []byte
}

func NewFieldEncryptor(hexKey string) (*FieldEncryptor, error) {
	if hexKey == "" {
		// Default 32-byte (256-bit) key for local dev
		hexKey = "6361666562616265313233343536373839303132333435363738393031323334"
	}

	key, err := hex.DecodeString(hexKey)
	if err != nil || len(key) != 32 {
		return nil, errors.New("encryption key must be a valid 64-character hex string (32 bytes / 256 bits)")
	}

	return &FieldEncryptor{secretKey: key}, nil
}

func (e *FieldEncryptor) Encrypt(plaintext string) (string, error) {
	if plaintext == "" {
		return "", nil
	}

	block, err := aes.NewCipher(e.secretKey)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}

	ciphertext := gcm.Seal(nonce, nonce, []byte(plaintext), nil)
	return hex.EncodeToString(ciphertext), nil
}

func (e *FieldEncryptor) Decrypt(ciphertextHex string) (string, error) {
	if ciphertextHex == "" {
		return "", nil
	}

	ciphertext, err := hex.DecodeString(ciphertextHex)
	if err != nil {
		return "", fmt.Errorf("invalid ciphertext hex: %w", err)
	}

	block, err := aes.NewCipher(e.secretKey)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonceSize := gcm.NonceSize()
	if len(ciphertext) < nonceSize {
		return "", errors.New("ciphertext too short")
	}

	nonce, ciphertextBytes := ciphertext[:nonceSize], ciphertext[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertextBytes, nil)
	if err != nil {
		return "", fmt.Errorf("decryption authentication failed: %w", err)
	}

	return string(plaintext), nil
}

File: python_backend/security/encryption.py
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

class FieldEncryptor:
    def __init__(self, hex_key=None):
        if not hex_key:
            hex_key = "6361666562616265313233343536373839303132333435363738393031323334"
            
        try:
            self.secret_key = bytes.fromhex(hex_key)
            if len(self.secret_key) != 32:
                raise ValueError
        except ValueError:
            raise ValueError("encryption key must be a valid 64-character hex string (32 bytes / 256 bits)")
            
    def encrypt(self, plaintext: str) -> str:
        if not plaintext:
            return ""
            
        aesgcm = AESGCM(self.secret_key)
        nonce = os.urandom(12)
        ciphertext = aesgcm.encrypt(nonce, plaintext.encode('utf-8'), None)
        return (nonce + ciphertext).hex()

    def decrypt(self, ciphertext_hex: str) -> str:
        if not ciphertext_hex:
            return ""
            
        try:
            data = bytes.fromhex(ciphertext_hex)
        except ValueError as e:
            raise ValueError(f"invalid ciphertext hex: {e}")
            
        if len(data) < 12:
            raise ValueError("ciphertext too short")
            
        nonce = data[:12]
        ciphertext = data[12:]
        
        aesgcm = AESGCM(self.secret_key)
        try:
            plaintext = aesgcm.decrypt(nonce, ciphertext, None)
        except Exception as e:
            raise ValueError(f"decryption authentication failed: {e}")
            
        return plaintext.decode('utf-8')
